### ISSUE TITLE
Some contacts fail to sync to Salesforce

### ISSUE THREAD

**Issue by davisagli** [washingtontrails/vms#477]
For Plone user accounts that existed before the switch to email-based logins, it's possible to have multiple contacts with Website Member Ids that are the same except for the capitalization. For example, we have `marco`, `Marco`, and `marcO`:
https://wta.my.salesforce.com/003A0000013w4cGIAQ
https://wta.my.salesforce.com/003A000001Jld0WIAR
https://wta.my.salesforce.com/003A000001cYj3RIAS

Also, the Website Member Id field does not enforce uniqueness, and it appears that at least in some cases there is more than one contact with the same value for Website Member Id. For example, these two are both `marco`:
https://wta.my.salesforce.com/003A0000013w4cGIAQ
https://wta.my.salesforce.com/003A000001boxByIAI

Both of these cause problems when one of these users saves updates to their profile, because the code that tries to write those updates to Salesforce gets a "SalesforceMoreThanOneRecord" error when it tries to match an existing contact using Website Member Id as an external id field.

I think the fix for the first issue is pretty easy; if we check "Case Sensitive" for the Website Member Id field at https://cs9.salesforce.com/00NA0000008qj0M?setupid=ContactFields, then it should consider variations in case to be separate values, which is what we need to support the legacy values from Plone.

To fix the second one we probably need to start by identifying how many of these there are, then do some manual cleanup to make sure there's only one contact for each website member id, then mark the field as unique.

/cc @mscholtz

**Comment by mscholtz** [comment]
That "Case Sensitive" setting, as I understand it, primarily applies to how SF imposes the uniqueness validation, assuming you have that turned on.  (The case-sensitivity is a modifier to the Unique setting)  It does also apply to how it does matching on upserts, but AFAIK you can't have the latter without the former

https://help.salesforce.com/apex/HTViewHelpDoc?id=faq_import_general_what_is_an_external.htm&language=en

That means that we would need to fix all examples of case sensitive duplicates in the data (like your last pair of examples) before turning this on.  Which is probably something that we should do regardless..

**Comment by davisagli** [comment]
Got it, I didn't see that you can only change the Case Sensitive setting when Unique is checked.

**Comment by davisagli** [comment]
Assigning to Charlie to set priority.

**Comment by WaBirder57** [comment]
Sounds like this requires a bit of clean up on the part of WTA before we can implement.  Is that correct, or can we do step one first and then later do the second step.

**Comment by mscholtz** [comment]
It needs some cleanup.  You could pull a report to find duplicate
usernames, and then resolve them via deletion, edit, or merge.  Or use
DemandTools etc.

There aren't currently SF IDs stored in plone, right, David?   So if they
merge contacts they don't have to be too careful about which ID they keep?

On Mon, Dec 7, 2015 at 3:12 PM, Charles Kahle <notifications@github.com>
wrote:

> Sounds like this requires a bit of clean up on the part of WTA before we
> can implement. Is that correct, or can we do step one first and then later
> do the second step.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/washingtontrails/vms/issues/477#issuecomment-162701192>
> .
>



-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by davisagli** [comment]
That's right, no SF ids stored in Plone. They do want to be careful to keep the id on the contact in SF that actually corresponds to the user in Plone with that id, though, if the contacts are not all actually the same person.

**Comment by davisagli** [comment]
I think it's actually fairly urgent that we get this sorted out. The problem is that if a user updates their profile and the changes fail to sync to Salesforce because of this issue, then their changes will get wiped out when contact values are synced from Salesforce to Plone overnight.

I'm attaching a CSV listing the contacts that have duplicate Website Member IDs. These are exact matches (it doesn't include the ones that are the same except for having different capitalization). Once these are not duplicated we can mark the Website Member ID field as unique and case sensitive, and then the same-except-capitalization ones will be able to sync.

[contacts_with_duplicate_member_ids.csv.txt](https://github.com/washingtontrails/vms/files/57332/contacts_with_duplicate_member_ids.csv.txt)

**Comment by davisagli** [comment]
Edited this to delete the attachment (I forgot this is a public tracker). Will send by email instead.

**Comment by WaBirder57** [comment]
Confirm for me the name of the Plone field that the Salesforce maps to.  Is it the Plone UserID field?

**Comment by davisagli** [comment]
Website Member ID (Website_Member_ID__c)

David Glick
(mobile)

**Comment by WaBirder57** [comment]
Is it possible or even advisable to stop the Plone <> SF Sync until we get this sorted out?

**Comment by WaBirder57** [comment]
One other question, can we make that field in SF - Webiste Member ID Case Sensitive so that values of "Adrian" and "adrian" would be considered different?  I assume not but wanted to double check to see if this was an option.

**Comment by davisagli** [comment]
@WaBirder57 we can make it Case Sensitive, but not until we make it Unique. And we can't make it Unique while there are duplicate values (with the same case). That is why I have asked you to focus on cleaning up the same-case duplicates.

**Comment by davisagli** [comment]
I don't think it makes sense to stop the updates from Plone to Salesforce. They are working for users who have non-duplicated member ids, so stopping it would just prevent more data from making it into Salesforce.

We might want to stop the nightly sync from Salesforce to Plone until we have this resolved. That would avoid overwriting new values entered in Plone with old values from Salesforce. A downside to this is that updates to member status would not make it to Plone. Also, those values would still get overwritten on users whose sync was failing (i.e. users with duplicated member ids) once we turn the sync back on.

**Comment by davisagli** [comment]
Charlie would it be helpful for me to export some of the data for the Plone users with those duplicated member ids so you can determine which contact is the right one to keep the member id? Or do you already have access to that in a way that works for you?

**Comment by WaBirder57** [comment]
I have access to the Plone Member Data dump that is run nightly.  This returns some fields for all the Plone accounts.  If you could just export those that are duplicated which we need to focus on that would be a big help.

**Comment by davisagli** [comment]
What fields do you want?

**Comment by WaBirder57** [comment]
The fields that I am aware of that would help us ID the plone users in Salesforce are below.  If you know of any other fields that you think might help add those also.  Thanks,.

UserID
Email
ScreenName
FirstName
LastName
PostalCode
MostRecentLogin
PreviousLogin
MostRecentTripReportTitle

**Comment by WaBirder57** [comment]
What happens if we go through this process and find the Salesforce Contacts with the wrong Plone UserID and just delete the value from Salesforce.  Will the Plone sync to Salesforce be able to find that existing SF record that matches based on the criteria we have defined and populate that existing SF contact with the correct Plone User ID?  Same methodology if a new MyBackpack is created and we check for an existing SF Contact before a CF Contact is created.

**Comment by davisagli** [comment]
No, that won't quite work. Plone keeps track of which users have been connected to a Salesforce contact. When a user logs in who has not been connected, it calls the vmslogin webservice to match an existing contact (first by user id -> Website Member ID, then by name and email) or create a new one. But if a user logs in who Plone knows was previously connected, this does not happen. In this case if the user updates their profile and there is no longer a contact in Salesforce with a matching Website Member Id, the sync to SF will fail.

However I could pretty easily reset the "connected to Salesforce" flag to False for the users whose user ids are duplicated. Then it'd be pretty safe to just remove the Website Member Id from all the contacts with one of the duplicated ids, and they'd get matched the next time the user logs in as long as there is a single contact with matching name and email.

**Comment by WaBirder57** [comment]
FYI I went and looked at our Nov 9th Salesforce backup, our Contacts table had the same 2 Contacts with the WebsiteMemberID value of "marco".  So I think this is good in that it indicates duplicates existed before we starting using MBP.  Thus, not a problem introduced by MBP.

**Comment by jessesnyder** [comment]
There were two unrelated bugs in the Plone system that were clouding our troubleshooting/diagnostic efforts yesterday:

* A homepage URL with the format 'http://[url=https://foo.com/bar/]https://foo.com/bar/[/url]' was breaking the member data dump CSV report, so I was looking repeatedly at a report from 11/22 when I thought I was looking at a fresh copy
* The registry that translates between user IDs and screen names was not removing old screenname-to-userID mappings when a user's screenname was updated, so using the old screenname in a backpack URL still worked (this is why the url http://www.wta.org/@@backpacks/scrnm-sparkymarc still directs you to the backpack of the userD "marco", even though that user's current screenname is "marco", not "sparkymarc")

I've fixed both bugs, but have not yet been able to rebuild the screenname<-->userId registry. A recent Plone hotfix prevents me from running the code to do that until I update it to be compliant with the new tighter security practices.

**Comment by WaBirder57** [comment]
@davisagli We had a question on the best way to clean up these SF Contacts which have duplicate Website Member ID.  Could we just remove these values from all the affected Contact records and then have Plone code that would make a single pass to try and match these to a plone account?

This would leave us with SF Contacts where all the Website User ID are unique so we can change this to field to be Unique and Case Sensitive.

Also we would have SF Contacts with a value for Website Member Name but no Website Member ID.  That would be our list to track down and try and match to existing Plone accounts but that would not hold up the rest of this work.

**Comment by davisagli** [comment]
@WaBirder57 Could we just remove these values from all the affected Contact records and then have Plone code that would make a single pass to try and match these to a plone account?

I could pretty easily write a script that calls the vmslogin webservice to match a contact for each of these users. However I'm not sure what the webservice does if it finds multiple contacts with a matching name and email address. @mscholtz?

> Also we would have SF Contacts with a value for Website Member Name but no Website Member ID. That would be our list to track down and try and match to existing Plone accounts but that would not hold up the rest of this work.

I think it would make sense to remove the Website Member Name from the same records from which you remove the Website Member Id...the field is meant to be a copy of the screen name for the matching Plone user, so it should be blank if the plone user is unknown, right?

**Comment by WaBirder57** [comment]
For now leaving the unmatched Contact.WebsiteMemberName populated is helpful for us as we try and track down the correct Plone account.  This has already helped us once.  Only when we are finished this process of clean up should we can delete those values.

**Comment by mscholtz** [comment]
Charlie I think this would only work if every contact to be matched in sf
has a unique name and email, which seems like an even taller order than
having unique Plone ID.

On Thursday, December 10, 2015, Charles Kahle <notifications@github.com>
wrote:

> For now leaving the unmatched Contact.WebsiteMemberName populated is
> helpful for us as we try and track down the correct Plone account. This has
> already helped us once. Only when we are finished this process of clean up
> should we can delete those values.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/washingtontrails/vms/issues/477#issuecomment-163802773>
> .
>


-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by davisagli** [comment]
To answer my own question, it looks like if there are multiple contacts matching name and email, the vmslogin webservice (well actually the WTA_VMS_ContactFinderCreator helper class) will use the first one (but the order is not defined, so basically it'll pick one of them randomly).

**Comment by mscholtz** [comment]
Yep. We actually figured this out at the meeting earlier.  I think the
original plan is better.  Correct the IDs don't just blank them all out.

On Thursday, December 10, 2015, David Glick <notifications@github.com>
wrote:

> To answer my own question, it looks like if there are multiple contacts
> matching name and email, the vmslogin webservice (well actually the
> WTA_VMS_ContactFinderCreator helper class) will use the first one (but the
> order is not defined, so basically it'll pick one of them randomly).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/washingtontrails/vms/issues/477#issuecomment-163812400>
> .
>


-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by davisagli** [comment]
> Charlie I think this would only work if every contact to be matched in sf
has a unique name and email, which seems like an even taller order than
having unique Plone ID.

Actually I just checked, and of the contacts who have the duplicated member ids, only 1 of the member ids belongs to multiple contacts with the exact same name and email. Even better, that's because the email is empty and the member id is vedaw who was a former Groundwire employee, so those contacts are probably left over from testing long ago.

So: I think Charlie's approach may actually be a good one, to first wipe out the member ids that are duplicated and then run a script to force re-matching the plone users with those ids.

However: We're assuming that the Plone users have the correct name and email, and that may not be the case since the nightly sync from SF has overwritten them with data from (randomly) one of the multiple contacts with a matching Website Member Id. Right? So before we do the above we might want to reset at least the name and email of those Plone users to what they were before we launched MBP, if we have a memberdata dump from then. Or do you have another plan for dealing with that problem?

**Comment by mscholtz** [comment]
That's right and the plan to deal with that was what I already put
forward.  If Charlie leaves the member ID with the contact who should have
it then it will correct any wrong data in Plone when it syncs.

On Thursday, December 10, 2015, David Glick <notifications@github.com>
wrote:

> Charlie I think this would only work if every contact to be matched in sf
> has a unique name and email, which seems like an even taller order than
> having unique Plone ID.
>
> Actually I just checked, and of the contacts who have the duplicated
> member ids, only 1 of the member ids belongs to multiple contacts with the
> exact same name and email. Even better, that's because the email is empty
> and the member id is vedaw who was a former Groundwire employee, so those
> contacts are probably left over from testing long ago.
>
> So: I think Charlie's approach may actually be a good one, to first wipe
> out the member ids that are duplicated and then run a script to force
> re-matching the plone users with those ids.
>
> However: We're assuming that the Plone users have the correct name and
> email, and that may not be the case since the nightly sync from SF has
> overwritten them with data from (randomly) one of the multiple contacts
> with a matching Website Member Id. Right? So before we do the above we
> might want to reset at least the name and email of those Plone users to
> what they were before we launched MBP, if we have a memberdata dump from
> then. Or do you have another plan for dealing with that problem?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/washingtontrails/vms/issues/477#issuecomment-163847858>
> .
>


-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by WaBirder57** [comment]
Thanks for all the thoughts on this.  We are going to figure this out on
Tuesday when I get back.  For now, staff is working through the data buy
hand and fixing the high priority ones (eg members and users with many
assets in their my back pack).

Arron is making another run with Demand Tools to dedup Salesforce contacts,
especially those with same namd and email (he said he found 18 of those).

Jesse send a Plone data dump from Aug 18 for use to use also.

Charlie

Charlie Kahle
Information Services Manager
Washington Trails Association
705 Second Ave, Suite 300
Seattle, Wa 98104
206.625.1367 - office
206.388.4511 - direct
www.wta.org

**Comment by mscholtz** [comment]
OK.  This doesn't affect the other SF data migration, right?  I can still
go ahead on Monday and create WPs and WP Registrations with the contacts as
they stand now?

(I don't see any problem there, just making sure no-one else does)

On Fri, Dec 11, 2015 at 9:03 AM, Charles Kahle <notifications@github.com>
wrote:

> Thanks for all the thoughts on this. We are going to figure this out on
> Tuesday when I get back. For now, staff is working through the data buy
> hand and fixing the high priority ones (eg members and users with many
> assets in their my back pack).
>
> Arron is making another run with Demand Tools to dedup Salesforce contacts,
> especially those with same namd and email (he said he found 18 of those).
>
> Jesse send a Plone data dump from Aug 18 for use to use also.
>
> Charlie
>
> Charlie Kahle
> Information Services Manager
> Washington Trails Association
> 705 Second Ave, Suite 300
> Seattle, Wa 98104
> 206.625.1367 - office
> 206.388.4511 - direct
> www.wta.org
>
> On Thu, Dec 10, 2015 at 10:39 PM, Matthew Scholtz <
> notifications@github.com>
> wrote:
>
> > That's right and the plan to deal with that was what I already put
> > forward. If Charlie leaves the member ID with the contact who should have
> > it then it will correct any wrong data in Plone when it syncs.
> >
> > On Thursday, December 10, 2015, David Glick <notifications@github.com>
> > wrote:
> >
> > > Charlie I think this would only work if every contact to be matched in
> sf
> > > has a unique name and email, which seems like an even taller order than
> > > having unique Plone ID.
> > >
> > > Actually I just checked, and of the contacts who have the duplicated
> > > member ids, only 1 of the member ids belongs to multiple contacts with
> > the
> > > exact same name and email. Even better, that's because the email is
> empty
> > > and the member id is vedaw who was a former Groundwire employee, so
> those
> > > contacts are probably left over from testing long ago.
> > >
> > > So: I think Charlie's approach may actually be a good one, to first
> wipe
> > > out the member ids that are duplicated and then run a script to force
> > > re-matching the plone users with those ids.
> > >
> > > However: We're assuming that the Plone users have the correct name and
> > > email, and that may not be the case since the nightly sync from SF has
> > > overwritten them with data from (randomly) one of the multiple contacts
> > > with a matching Website Member Id. Right? So before we do the above we
> > > might want to reset at least the name and email of those Plone users to
> > > what they were before we launched MBP, if we have a memberdata dump
> from
> > > then. Or do you have another plan for dealing with that problem?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
> https://github.com/washingtontrails/vms/issues/477#issuecomment-163847858>
> > > .
> > >
> >
> >
> > --
> >
> > Matthew Scholtz
> > Percolator Consulting
> > (206) 289-0816 x3
> >
> > http://percolatorconsulting.com/
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/washingtontrails/vms/issues/477#issuecomment-163856619>
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/washingtontrails/vms/issues/477#issuecomment-163991185>
> .
>



-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by davisagli** [comment]
@WaBirder57 sounds good. Please let me know if/when you anticipate wanting to run that script to force re-matching Plone users, so I can make sure I'm ready to do that.

**Comment by davisagli** [comment]
@WaBirder57 any update on what you are planning to do with these duplicate member ids? There are still occasional updates failing to sync to Salesforce (2 so far today).

**Comment by WaBirder57** [comment]
We are continuing to work through the list of My Backpack accounts that link to more than one Salesforce contact.  We are finding most often that the SF Contact has a duplicate or that 2 contacts in one Household are both linked to the same MyBackpack account.  We have prioritized those users who use MyBackpack the most.  Still a couple weeks of work to get this cleaned up unless we need to bump this to a higher priority task.

**Comment by WaBirder57** [comment]
I changed the priority to Medium as this doesn't need to be fixed before VMS goes live.  We are slowly working through these and have complete those users who seem most active users of My Backpack so their info will sycn with Salesforce.
